geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/function/BinaryOutputEncoder.scala

```diff
@@ -0,0 +1,229 @@
+    // depending on srs requested and wfs versions, axis order can be flipped
+    val getLineLatLon: (SimpleFeature) => Array[(Float, Float)] = axisOrder match {
+      case LatLon => (f) => {
+        val geom = f.getDefaultGeometry.asInstanceOf[LineString]
+        val points = (0 until geom.getNumPoints).map(geom.getPointN(_))
+        points.map(p => (p.getX.toFloat, p.getY.toFloat)).toArray
+      }
+      case LonLat => (f) => {
+        val geom = f.getDefaultGeometry.asInstanceOf[LineString]
+        val points = (0 until geom.getNumPoints).map(geom.getPointN(_))
+        points.map(p => (p.getY.toFloat, p.getX.toFloat)).toArray
+      }
+    }
+
+    // gets the track id from a feature
+    val getTrackId: (SimpleFeature) => Option[String] = trackIdField match {
+      case Some(trackId) if (trackId == "id") => (f) => Some(f.getID)
+
+      case Some(trackId) =>
+        val trackIndex  = sft.indexOf(trackId)
+        (f) => Option(f.getAttribute(trackIndex)).map(_.toString)
+
+      case None => (_) => None
+    }
+
+    // gets the label from a feature
+    val getLabel: (SimpleFeature) => Option[String] = labelField match {
+      case Some(label) if (label == "id") => (f) => Some(f.getID)
+      case Some(label) =>
+        val labelIndex = sft.indexOf(label)
+        (f) => Option(f.getAttribute(labelIndex)).map(_.toString)
+      case None => (_) => None
+    }
+
+    // encodes the values in either 16 or 24 bytes
+    val encode: (ValuesToEncode) => Unit = labelField match {
+      case Some(label) => (v) => {
+        val toEncode = ExtendedValues(v.lat, v.lon, v.dtg, v.track, v.label)
+        Convert2ViewerFunction.encode(toEncode, output)
+      }
+      case None => (v) => {
+        val toEncode = BasicValues(v.lat, v.lon, v.dtg, v.track)
+        Convert2ViewerFunction.encode(toEncode, output)
+      }
+    }
+
+    val iter = if (isLineString) {
+      // expand the line string into individual points to encode
+      new RichSimpleFeatureIterator(fc.features()).flatMap { sf =>
+        val points = getLineLatLon(sf)
+        val dates = getLineDtg(sf)
+        if (points.size != dates.size) {
+          logger.warn(s"Mismatched geometries and dates for simple feature ${sf.getID} - skipping")
+          Seq.empty
+        } else {
+          val trackId = getTrackId(sf)
+          val label = getLabel(sf)
+          (0 until points.size).map { case i =>
+            val (lat, lon) = points(i)
+            ValuesToEncode(lat, lon, dates(i), trackId, label)
+          }
+        }
+      }
+    } else {
+      new RichSimpleFeatureIterator(fc.features()).map { sf =>
+        val (lat, lon) = getLatLon(sf)
+        val dtg = getDtg(sf)
+        val trackId = getTrackId(sf)
+        val label = getLabel(sf)
+        ValuesToEncode(lat, lon, dtg, trackId, label)
+      }
+    }
+    if (sort) {
+      val set = new scala.collection.mutable.TreeSet[ValuesToEncode]()
+      iter.foreach(set.add)
+      set.foreach(encode)
```

Comment:
Probably not... http://stackoverflow.com/questions/3607593/is-it-faster-to-add-to-a-collection-then-sort-it-or-add-to-a-sorted-collection
I'll change.